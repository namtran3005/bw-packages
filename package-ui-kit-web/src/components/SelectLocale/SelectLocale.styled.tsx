import * as React from 'react';
import styled from 'styled-components';
import { ButtonBase as MuiButtonBase } from '@material-ui/core';
import { ButtonBaseProps as MuiButtonBaseProps } from '@material-ui/core/ButtonBase';
import { ColorsType, Colors, Spacing } from '@bitwala-cryptobank-squad/package-theme';
import { Paper, PaperProps } from '../Paper';
import { List, ListProps, ListItem, ListItemProps } from '../List';

interface LocaleIconButtonProps extends MuiButtonBaseProps {
  background?: ColorsType;
  foreground?: ColorsType;
  ripple?: boolean;
  hoverForeground?: ColorsType;
}

interface SelectLocaleMenuProps {
  paperProps?: PaperProps;
  listProps?: ListProps;
  dropdownHeight?: 'small' | 'medium' | 'big';
}

interface SelectLocaleMenuItemProps extends ListItemProps {
  selected?: boolean;
  highlighted?: boolean;
}

const SelectLocaleMenuItemComponent: React.FunctionComponent<SelectLocaleMenuItemProps> = ({
  /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
  selected,
  /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
  highlighted,
  ...cleanProps
}) => <ListItem {...cleanProps} />;

const SelectLocaleMenuComponent: React.FunctionComponent<SelectLocaleMenuProps> = ({
  paperProps,
  listProps,
  children,
  /* eslint-disable-next-line @typescript-eslint/no-unused-vars */
  dropdownHeight,
  ...cleanProps
}) => (
  <Paper spacing={0} elevation={24} {...paperProps} {...cleanProps}>
    <List {...listProps}>{children}</List>
  </Paper>
);

const LocaleIconButtonComponent: React.FunctionComponent<LocaleIconButtonProps> = ({
  /* eslint-disable @typescript-eslint/no-unused-vars */
  background,
  foreground,
  hoverForeground,
  /* eslint-enable @typescript-eslint/no-unused-vars */
  ripple = true,
  ...cleanProps
}) => (
  <MuiButtonBase
    focusRipple={!ripple}
    disableRipple={!ripple}
    {...cleanProps}
  />
);

const SelectLocaleMenuItem = styled(SelectLocaleMenuItemComponent)`
  &&& {
    background-color: ${({ highlighted, selected }) =>
      highlighted || selected
        ? Colors.get('primaryGrey')
        : Colors.getWithAlpha('white', 0)} !important;
    cursor: pointer;

    &:hover {
      box-shadow: none !important;
    }
  }
`;

const SelectLocaleMenu = styled(SelectLocaleMenuComponent)`
  &&& {
    position: absolute;
    top: 100%;
    right: 0;
    max-height: ${({ dropdownHeight = 'small' }) =>
      dropdownHeight === 'small'
        ? '140px'
        : dropdownHeight === 'medium'
        ? '243px'
        : dropdownHeight === 'big' && '294px'};
    z-index: 6230;
    overflow: auto;
  }
`;

const IconContainer = styled.span`
  display: inline-block;
  margin: 0 ${Spacing.unit}px;
`;

const ChevronDown = styled.span`
  &&& {
    display: inline-block;
    font-size: 0;
    line-height: 0;
    transform: rotate(90deg);
  }
`;

const LocaleIconButton = styled(LocaleIconButtonComponent)`
  &&& {
    white-space: nowrap;
    border-radius: 0;
    line-height: 0;
    background: ${({ background }) =>
      background ? Colors.get(background) : 'inherit'};
    color: ${({ foreground }) =>
      foreground ? Colors.get(foreground) : 'inherit'};
    display: inline-block;
    padding: ${({ ripple = true }) =>
      ripple ? `${Spacing.unit}px ${Spacing.unit * 1.5}px` : '0px'};

    ${({ ripple = true }) =>
      ripple
        ? null
        : `
      background: transparent;
      opacity: 0.5;
    `};

    &:hover {
      color: ${({ hoverForeground }) =>
        Colors.get(hoverForeground ? hoverForeground : 'primaryBlack')};
    }

    &:focus {
      ${({ ripple = true }) => (ripple ? null : 'opacity: 0.8;')};
    }

    &:active {
      ${({ ripple = true }) => (ripple ? null : 'opacity: 1;')};
    }
  }
`;

const Root = styled.div`
  &&& {
    position: relative;
  }
`;

Root.displayName = 'SelectLocale';

export {
  Root,
  ChevronDown,
  IconContainer,
  LocaleIconButton,
  LocaleIconButtonProps,
  LocaleIconButtonComponent,
  SelectLocaleMenu,
  SelectLocaleMenuProps,
  SelectLocaleMenuComponent,
  SelectLocaleMenuItem,
  SelectLocaleMenuItemProps,
  SelectLocaleMenuItemComponent,
};
