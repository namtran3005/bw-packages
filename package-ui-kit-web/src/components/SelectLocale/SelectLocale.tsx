import React, { useState } from 'react';
import Downshift, {
  DownshiftProps,
  ControllerStateAndHelpers,
} from 'downshift';
import {
  ChevronRightIcon,
  DeFlagIcon,
  UkFlagIcon,
} from '@bitwala-cryptobank-squad/package-icons';
import {
  Locales,
  localeOptions as defaultLocaleOptions,
} from '@bitwala-cryptobank-squad/package-constants';

import { Icon } from '../Icon';
import { ListItemIcon } from '../List';
import { Typography } from '../Typography/';
import {
  Root,
  ChevronDown,
  IconContainer,
  LocaleIconButton,
  LocaleIconButtonProps,
  SelectLocaleMenu,
  SelectLocaleMenuItem,
  SelectLocaleMenuProps,
} from './SelectLocale.styled';

interface LocaleIconsMap {
  [Locales.en]: React.ReactNode;
  [Locales.de]: React.ReactNode;
}

interface LocaleOptionType {
  label: string;
  value: string;
}

interface SelectLocaleProps {
  downshiftProps?: DownshiftProps<LocaleOptionType>;
  dropdownHeight?: SelectLocaleMenuProps['dropdownHeight'];
  getLocaleIcon: (localeOptionValue: string) => React.ReactNode;
  localeButtonProps?: LocaleIconButtonProps;
  localeOptions: LocaleOptionType[];
  onChange?: (
    selectedItem: LocaleOptionType,
    stateAndHelpers: ControllerStateAndHelpers<LocaleOptionType>
  ) => void;
  value: string;
}

const localeIconsMap: LocaleIconsMap = {
  [Locales.en]: <UkFlagIcon />,
  [Locales.de]: <DeFlagIcon />,
};

const defaultGetLocaleIcon = (locale: Locales): React.ReactNode =>
  localeIconsMap[locale];

const itemToString = (item: LocaleOptionType) => {
  if (item) {
    return item.label || item.value;
  }
  return '';
};

function SelectLocale(props: SelectLocaleProps) {
  const [isMenuOpen, setMenuState] = useState(false);

  const {
    downshiftProps,
    dropdownHeight,
    getLocaleIcon,
    localeButtonProps,
    localeOptions,
    value,
  } = props;

  const openMenu = () => setMenuState(true);
  const closeMenu = () => setMenuState(false);

  const getInitialSelectedItem = (selectedValue: string) => {
    const foundOption = localeOptions.find(
      (localeOption) => localeOption.value === selectedValue
    );
    if (foundOption) {
      return foundOption;
    }
    return null;
  };

  const handleSelect = (
    selectedItem: LocaleOptionType,
    stateAndHelpers: ControllerStateAndHelpers<LocaleOptionType>
  ) => {
    if (props.onChange) {
      props.onChange(selectedItem, stateAndHelpers);
    }

    closeMenu();
  };

  const isSelected = (
    selectedItem: LocaleOptionType,
    item: LocaleOptionType
  ): boolean => {
    if (selectedItem) {
      return selectedItem.value === item.value;
    }
    return false;
  };

  return (
    <Downshift
      itemToString={itemToString}
      isOpen={isMenuOpen}
      onOuterClick={closeMenu}
      onSelect={handleSelect}
      initialSelectedItem={getInitialSelectedItem(value)}
      {...downshiftProps}
    >
      {({
        getItemProps,
        getMenuProps,
        isOpen,
        highlightedIndex,
        selectedItem,
      }) => (
        <div>
          <Root data-test="locale-toggle">
            <div>
              <LocaleIconButton
                classes={{
                  disabled: 'disabled',
                }}
                {...localeButtonProps}
                onClick={isOpen ? closeMenu : openMenu}
              >
                <IconContainer>
                  <Icon size={24}>{getLocaleIcon(value)}</Icon>
                </IconContainer>
                <ChevronDown>
                  <ChevronRightIcon />
                </ChevronDown>
              </LocaleIconButton>
            </div>
            {isOpen && (
              <SelectLocaleMenu
                dropdownHeight={dropdownHeight}
                listProps={{ ...getMenuProps() }}
              >
                {localeOptions.length > 0 &&
                  localeOptions.map((item, index) => (
                    <SelectLocaleMenuItem
                      key={`locale_${index}`}
                      selected={isSelected(selectedItem, item)}
                      highlighted={highlightedIndex === index}
                      {...getItemProps({
                        key: item.value,
                        index,
                        item,
                      })}
                    >
                      <ListItemIcon key={index}>
                        <Icon size={24}>{getLocaleIcon(item.value)}</Icon>
                      </ListItemIcon>
                      <Typography>{item.label || item.value}</Typography>
                    </SelectLocaleMenuItem>
                  ))}
              </SelectLocaleMenu>
            )}
          </Root>
        </div>
      )}
    </Downshift>
  );
}

SelectLocale.defaultProps = {
  getLocaleIcon: defaultGetLocaleIcon,
  localeOptions: defaultLocaleOptions,
  value: Locales.en,
};

export { SelectLocale, SelectLocaleProps, LocaleOptionType };
