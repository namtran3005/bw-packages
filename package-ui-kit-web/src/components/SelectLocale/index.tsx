export {
  SelectLocale,
  SelectLocaleProps,
  LocaleOptionType,
} from './SelectLocale';
