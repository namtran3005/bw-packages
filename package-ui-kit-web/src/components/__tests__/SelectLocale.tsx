import * as React from 'react';
import { shallow } from 'enzyme';
import { SelectLocale } from '../SelectLocale';

const defaultProps = {
  // eslint-disable-next-line @typescript-eslint/prefer-as-const
  dropdownHeight: 'small' as 'small',
  localeOptions: [
    {
      value: 'opt1',
      label: 'Option 1',
    },
    {
      value: 'opt2',
      label: 'Option 2',
    },
  ],
};

describe('Autocomplete', () => {
  it('should render submit Autocomplete', () => {
    const wrapper = shallow(<SelectLocale {...defaultProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should render submit Autocomplete with preselected options', () => {
    const wrapper = shallow(<SelectLocale {...defaultProps} value="opt2" />);
    expect(wrapper).toMatchSnapshot();
  });
});
