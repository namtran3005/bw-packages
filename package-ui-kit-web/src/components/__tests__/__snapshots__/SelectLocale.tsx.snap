// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete should render submit Autocomplete 1`] = `
<Downshift
  defaultHighlightedIndex={null}
  defaultIsOpen={false}
  environment={[Window]}
  getA11yStatusMessage={[Function]}
  initialSelectedItem={null}
  isOpen={false}
  itemToString={[Function]}
  onChange={[Function]}
  onInputValueChange={[Function]}
  onOuterClick={[Function]}
  onSelect={[Function]}
  onStateChange={[Function]}
  onUserAction={[Function]}
  scrollIntoView={[Function]}
  selectedItemChanged={[Function]}
  stateReducer={[Function]}
  suppressRefError={false}
>
  <Component />
</Downshift>
`;

exports[`Autocomplete should render submit Autocomplete with preselected options 1`] = `
<Downshift
  defaultHighlightedIndex={null}
  defaultIsOpen={false}
  environment={[Window]}
  getA11yStatusMessage={[Function]}
  initialSelectedItem={
    Object {
      "label": "Option 2",
      "value": "opt2",
    }
  }
  isOpen={false}
  itemToString={[Function]}
  onChange={[Function]}
  onInputValueChange={[Function]}
  onOuterClick={[Function]}
  onSelect={[Function]}
  onStateChange={[Function]}
  onUserAction={[Function]}
  scrollIntoView={[Function]}
  selectedItemChanged={[Function]}
  stateReducer={[Function]}
  suppressRefError={false}
>
  <Component />
</Downshift>
`;
